overlord/snapstate/handlers.go

```diff
@@ -646,6 +704,10 @@ func (m *SnapManager) undoLinkSnap(t *state.Task, _ *tomb.Tomb) error {
 		return err
 	}
 
+	if err := m.removeSnapContext(st, snapsup.Name()); err != nil {
```

Comment:
Similarly, we want to do this only if what is being undone is the linking of the very first snap, instead of every revision. This is actually a bug rather than just an optimization.. as it is undoing an installation will kill the previous snap's context. Probably worth a spread test.